Double tap camera

claudds opened the pull request:
Implemented double tap functionality on quick attachment camera in order to flip camera direction. Note that this functionality only works on the quick camera (which is activated by the camera icon in the compose panel). If users choose the camera option in the attachments menu, this will activate the specific phone's native camera and this feature will not be available. 

Introduces an icon in order to demonstrate which side of the camera is currently being viewed (mainly for testing purposes). 

Closes #95

claudds committed:
#95 #132 #133 paired with @samantha-kerr. Implemented new Gesture detector to detect double taps on the quick attachment camera

claudds committed:
#95 #132 #133 refactored feature to the quick attachment drawer so that the camera direction icon flips when the camera direction

samantha-kerr committed:
#95 #132 #133 Extracted DoubleTapGestureDetector class to an external public util class

claudds committed:
#95 #133 #132 added icon to show front and rear camera states, mainly for testing purposes

claudds committed:
#95 #134 espresso test for double tapping the camera. Implemented using expectations and conversation actions classes

claudds commented:
Note: unit tests pending

benrs commented on test/androidTest/java/org/thoughtcrime/securesms/util/Expectations.java:
```diff
@@ -19,6 +21,10 @@ public static void checkIsDisplayed(int id) {
         onView(withId(id)).check(matches(isDisplayed()));
     }
 
+    public static void checkIsDisplayedWithRearCameraDrawable(int id){
```
Could make the naming more generic "checkIsDisplayedWithResource"

claudds replied:
This is actually a method that i'm no longer using. Will remove

claudds committed:
#95 fixing review comments

samantha-kerr committed:
#95 #134 #148 Unit tests for camera double tap feature.

- CameraViewUnitTest tests the existing flipCamera() method, asserting the toggling of the cameraId based on whether the camera faces back or front; also verifyies that the setDirectCaptureCameraId() method is called within the flipCamera() method.

- DoubleTapGestureDetectorUnitTest tests that a double tap event is captured; assert that onDoubleTapEvent() is true and onSingleTapConfirmed() is false in this case, and verifies that the flipCamera() method is invoked by the CameraView when the double tap event is triggered.

samantha-kerr committed:
Merge branch 'double-tap-camera' of https://github.com/jacobrs/Signal-Android into double-tap-camera

benrs reviewed (approved):
Looks good.

samantha-kerr committed:
#95 #134 #148 Added unit test for camera icon visibility when camera is flipped

samantha-kerr committed:
#95 #134 #148 Refactored CameraViewUnitTest, moved stubs before method calls, removed verification of setDirectCaptureCameraId due to need for excessive mocking and stubbing of dependencies.

samantha-kerr commented:
Added and refactored some unit tests

lori-dalkin reviewed (approved):
Looks good. It seems like you ended up adding in an icon. Could be useful to post a final design in order to see the new addition.

claudds commented:
mockups added to story!
